### DISCUSSION TITLE
LSP Remote actions

### DISCUSSION THREAD

**Discussion by rasulomaroff** [folke/flash.nvim#79] [Show and tell]
Before Neovim I've been using JetBrains products for a long period of time and there's a plugin called `Acejump` there. The thing is that besides jumping the plugin also offers "Go to definition on jump" and I find that feature very convenient.

I tried implementing the same with `flash`, moreover, it can do much more, and it looks like that:
```lua
keys = {
  {
    'gm',
    function()
      local prev_timeout = vim.opt.timeout

      vim.opt.timeout = false

      require('flash').jump {
        action = function(match, state)
          state:hide()

          local row, col = unpack(vim.api.nvim_win_get_cursor(0))

          vim.api.nvim_set_current_win(match.win)
          vim.api.nvim_win_set_cursor(match.win, match.pos)

          local key = vim.api.nvim_replace_termcodes('<Ignore>' .. 'g', true, true, true)

          vim.api.nvim_feedkeys(key, 'i', false)

          vim.schedule(function()
            vim.api.nvim_win_set_cursor(match.win, { row, col })
            vim.opt.timeout = prev_timeout
          end)
        end,
        search = {
          max_length = 2,
        },
        highlight = {
          label = {
            before = { 0, 2 },
            after = false,
          }
        }
      }
    end,
  }
} 
```

All my LSP action keybindings start with 'g':
'gt' - go to definition
'gT' - go to type definition
'gr' - references
'gd' - go to declaration
etc

So, instead of jumping and then performing the action, this keybinding combines them into one. This can be applied not only for LSP actions, just replace 'g' in `nvim_replace_termcodes` with your preferred key. I think you got the idea.
`max_length` and `label` options can also be turned off, I just like a fixed pattern so I left them.

That code will perform LSP action and return your cursor back (this can be turned off by just deleting the first line inside the `vim.schedule` callback). Restoring a cursor position using `flash.remote` or `remote_op` didn't work for me for some reason, maybe someone could help to find out why?.

By the way, I've been using Neovim for a bit more than a month, so there can be bad practices in code. If there are, please tell me so that I fixed them and learn more 😅

**Comment by folke** [comment]
That's a really cool idea!

The whole remote stuff has been integrated in regular jump, so I removed that other code.

You can use `state:restore()` inside a `vim.schedule`, or for more advanced restore, you can use `require("flash.jump").restore_remote(state)`.

But for your use-case a simple restore is all you need.

**Reply by rasulomaroff** [reply]
Yep, `state:restore()` works perfectly there. Thanks!
